File: frontend/lib/api/cases.ts
import { request } from "./core";
import type { CaseLawSearchItem } from "../api";

export type Case = {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  case_number: string | null;
  status: string;
  created_at: string;
  updated_at: string;
};

export type CaseDetail = Case & {
  documents: Array<{
    id: string;
    document_type: string;
    document_category: string;
    created_at: string;
  }>;
  forum_posts: Array<{
    id: string;
    title: string;
    created_at: string;
  }>;
  case_law_items: CaseLawSearchItem[];
};

export type CaseDecisionSyncResponse = {
  status: string;
  case_id: string;
  case_number: string;
  records_in: number;
  created: number;
  updated: number;
  total: number;
};

export async function getCases(token?: string, demoUser?: string): Promise<Case[]> {
  return request<Case[]>("/api/cases", { token, demoUser });
}

export async function getCase(caseId: string, token?: string, demoUser?: string): Promise<CaseDetail> {
  return request<CaseDetail>(`/api/cases/${caseId}`, { token, demoUser });
}

export async function syncCaseDecisions(
  caseId: string,
  token?: string,
  demoUser?: string,
): Promise<CaseDecisionSyncResponse> {
  return request<CaseDecisionSyncResponse>(`/api/cases/${caseId}/sync-decisions`, {
    method: "POST",
    token,
    demoUser,
  });
}

export async function createCase(payload: Partial<Case>, token?: string, demoUser?: string): Promise<Case> {
  return request<Case>("/api/cases", { method: "POST", body: payload, token, demoUser });
}

export async function updateCase(
  caseId: string,
  payload: Partial<Case>,
  token?: string,
  demoUser?: string,
): Promise<Case> {
  return request<Case>(`/api/cases/${caseId}`, { method: "PATCH", body: payload, token, demoUser });
}

export async function deleteCase(caseId: string, token?: string, demoUser?: string): Promise<{ status: string }> {
  return request<{ status: string }>(`/api/cases/${caseId}`, { method: "DELETE", token, demoUser });
}

File: frontend/app/dashboard/analyze/cases.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload
from uuid import UUID
from typing import List

from app.db import get_session
from app.models.case import Case, CaseCreate, CaseUpdate, CaseInDB, CaseDetail
from app.models.user import User
from app.services.auth import get_current_user

router = APIRouter()

@router.post("/", response_model=CaseInDB, status_code=status.HTTP_201_CREATED, summary="Create a new case")
async def create_case(
    case_in: CaseCreate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session)
):
    new_case = Case(
        user_id=current_user.id,
        title=case_in.title,
        description=case_in.description,
        case_number=case_in.case_number,
        status=case_in.status
    )
    session.add(new_case)
    await session.commit()
    await session.refresh(new_case)
    return new_case

@router.get("/", response_model=List[CaseInDB], summary="Get all cases for the current user")
async def get_cases(
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session)
):
    result = await session.execute(
        select(Case).filter(Case.user_id == current_user.id).order_by(Case.created_at.desc())
    )
    cases = result.scalars().all()
    return cases

@router.get("/{case_id}", response_model=CaseDetail, summary="Get a specific case by ID")
async def get_case(
    case_id: UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session)
):
    result = await session.execute(
        select(Case)
        .options(
            selectinload(Case.documents),
            selectinload(Case.forum_posts)
        ).filter(Case.id == case_id, Case.user_id == current_user.id)
    )
    case = result.scalars().first()
    if not case:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Case not found")
    return case

@router.patch("/{case_id}", response_model=CaseInDB, summary="Update an existing case")
async def update_case(
    case_id: UUID,
    case_update: CaseUpdate,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session)
):
    result = await session.execute(
        select(Case).filter(Case.id == case_id, Case.user_id == current_user.id)
    )
    case = result.scalars().first()
    if not case:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Case not found")

    update_data = case_update.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(case, key, value)
    
    session.add(case)
    await session.commit()
    await session.refresh(case)
    return case

@router.delete("/{case_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Delete a case")
async def delete_case(
    case_id: UUID,
    current_user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_session)
):
    result = await session.execute(
        select(Case).filter(Case.id == case_id, Case.user_id == current_user.id)
    )
    case = result.scalars().first()
    if not case:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Case not found")

    await session.delete(case)
    await session.commit()
    return